From: Modestas Vainius <modestas@vainius.eu>
Subject: [PATCH] Link MySQLe (libmysqld) to amarok executable.

When libmysqld.a is linked to amarok executable, it no longer needs
to be built with -fPIC. This avoid portability problems at low cost.
sqlcollection is essential Amarok plugin anyway.

Signed-off-by: Modestas Vainius <modestas@vainius.eu>

---
 src/CMakeLists.txt                          |   10 ++++++++++
 src/collection/sqlcollection/CMakeLists.txt |   17 +++--------------
 2 files changed, 13 insertions(+), 14 deletions(-)

diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index 68eb7fb..3861a9b 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -600,6 +600,16 @@ if(Q_WS_MAC)
 else(Q_WS_MAC)
     kde4_add_executable(amarok ${amarok_SRCS})
     target_link_libraries(amarok ${KDE4_KDEUI_LIBS} amaroklib )
+    # Link MySQLe to the executable. Hence -fPIC for libmysqld is not
+    # needed anymore.
+    target_link_libraries(amarok -Wl,-whole-archive ${MYSQL_EMBEDDED_LIBRARIES} -Wl,-no-whole-archive)
+    if(NOT WIN32)
+        if(NOT APPLE)
+            target_link_libraries(amarok crypt pthread dl wrap)
+        endif(NOT APPLE)
+    endif(NOT WIN32)
+    set_target_properties(amarok PROPERTIES LINK_FLAGS "-Wl,--allow-multiple-definition")
+
     install(TARGETS amarok ${INSTALL_TARGETS_DEFAULT_ARGS})
 endif(Q_WS_MAC)
 
diff --git a/src/collection/sqlcollection/CMakeLists.txt b/src/collection/sqlcollection/CMakeLists.txt
index fbfe450..fb750d5 100644
--- a/src/collection/sqlcollection/CMakeLists.txt
+++ b/src/collection/sqlcollection/CMakeLists.txt
@@ -56,28 +56,17 @@ SET(CMAKE_SHARED_LINKER_FLAGS ${CMAKE_SHARED_LINKER_FLAGS_NOFATALWARN} )
 STRING(REPLACE "-Wl,--fatal-warnings" "" CMAKE_MODULE_LINKER_FLAGS_NOFATALWARN ${CMAKE_MODULE_LINKER_FLAGS})
 SET(CMAKE_MODULE_LINKER_FLAGS ${CMAKE_MODULE_LINKER_FLAGS_NOFATALWARN} )
 
+STRING(REPLACE "-Wl,--no-undefined" "" CMAKE_MODULE_LINKER_FLAGS_ALLOW_UNDEFINED ${CMAKE_MODULE_LINKER_FLAGS})
+SET(CMAKE_MODULE_LINKER_FLAGS ${CMAKE_MODULE_LINKER_FLAGS_ALLOW_UNDEFINED} )
+
 target_link_libraries(amarok_collection-sqlcollection
     amaroklib
     amarokpud
     ${KDE4_KDEUI_LIBS}
     ${KDE4_KIO_LIBS}
     ${KDE4_THREADWEAVER_LIBRARIES}
-    ${MYSQL_EMBEDDED_LIBRARIES}
-    #this line may be needed on Debian or other systems where mysql is compiled
-    #in the prescence of yassl. However it seems to cause problems for other people,
-    #so its been disabled.
-    #See http://bugs.mysql.com/bug.php?id=21489
-    ${MYSQL_LIBRARIES}
-    ${CMAKE_DL_LIBS}
-    ${ZLIB_LIBRARIES}
 )
 
-if(NOT WIN32)
-    if(NOT APPLE)
-        target_link_libraries(amarok_collection-sqlcollection crypt pthread)
-    endif(NOT APPLE)
-endif(NOT WIN32)
-
 if(APPLE)
 	SET_TARGET_PROPERTIES(amarok_collection-sqlcollection PROPERTIES LINK_FLAGS "-undefined dynamic_lookup")
 endif(APPLE)
-- 
tg: (1e4d33f..) debian/mysqle_link_to_amarok_executable_feature (depends on: debian/mysqle_linkage_fix)
